### DISCUSSION TITLE
More API Support for addons?

### DISCUSSION THREAD

**Discussion by ghost** [Taureon/aps-plus-plus#213] [Suggestions]
## It will be nice if we can customize things like in minecraft bukkit/spigot api (thats the only decent thing that I make so, forgive me if it sounds stupid)
making feature requests 

1. **join quit events, death events, level up events** etc, with preventdefault/cancel propogation mechanisms
2. apart from number 1, you can add more events in the event api so we can intercept it or process it??
3. make an api that **creates and load tanks in runtime** (it is cool and can be useful in circumstances)
   -theres a game called rocketeer that allows you to design tank in runtime, its another cool feature in this api too if you dont mind styles changing a bit
   -api can include functions that regenerate mockup data and inserting index appropriately so clients wont scream error or load the wrong tank image/names
4. api that **alters upgrade trees and tank data in run time, similar to no.2**
5. api to **completely control entities in games, and how to spawn it etc**? control health, time to live, etc 
    i know you can do it with the existing ones but you can make it simpler for devs?
6. host **multi gamemodes** in one server
7. more api support for addons, instead of just messing with configs and events
8. **custom gamemode support for add-ons, and also allow people to load custom maps in arras**
9. **map making system** that allows users to make their own maps and load into gamemodes
10. _i will come back when i can think of more_

**Comment by Trioplane** [comment]
You can already alter upgrade trees at runtime, You can also alter tank data at runtime too. But it is buggy rn.
We are planning to make a map making system via web gui.
